avaje-config/src/main/java/io/avaje/config/InitialLoader.java

```diff
@@ -188,12 +189,14 @@ private boolean loadTest() {
   }
 
   /**
-   * Load configuration defined by a <em>load.properties</em> entry in properties file.
+   * Recursively Load configuration defined by a <em>load.properties</em> entry in properties file.
    */
   private void loadViaIndirection() {
     String paths = loadContext.indirectLocation();
-    if (paths != null) {
+    if (paths != null && !loadProperties.contains(paths)) {
```

Comment:
I'm wondering if we just fail fast here.

that is `if (loadProperties.contains(paths))` happens you throw an `IllegalStateException`.

Here is an example of my concern.

application.properties:

```properties
# application.properties
load.properties=common.properties my-app.properties
```

```properties
# common.properties
blah=asdfasdf
load.properties=env:///
```

```properties
# my-app.properties
blah=I_should_be_overridden_by_env_but_will_not_now
load.properties=env:///
```

Ultimately I think the logic of trying to maintain a set of visited resources is probably too confusing and a simple integer counter passed in the recursion would be enough. If the integer exceeds some value like say 64 then we probably have a cycle or some other abuse.

I also didn't know for sure that you were doing a recursive call so in theory a stackoverflow would occur so maybe this change is not even worth it. Using the integer would just make the error happen faster and possibly clearer what happened.

Reply:
> that is if (loadProperties.contains(paths)) happens you throw an IllegalStateException.

That'll fail on the first recursion

> Here is an example of my concern.

Made a test and seems to behave as expected

Reply:
> load.properties=env:///

Is this just a bad example?  avaje-config already takes environment variables into account, it isn't clear to me what a `env:///` actually means here?  Maybe there is a better example of your concern but as a secondary question - what are examples of `env:///`? what is that trying to achieve?

Reply:
He's talking about that uri thing in the other pr, it seems he wants to implement special logic for it

Reply:
@rbygrave 

Ignore env. Replace with some normal path.

Reply:
Earlier I had thought @SentryMan change would not work as I assumed load properties was DFS traversal.

It is not but rather BFS which I’m now thinking is incorrect behavior.

@rbygrave was that always the desire for chaining?

I’ll edit this comment in the morning if it does not make sense.

Reply:
> I’m now thinking is incorrect behavior.

What should it be?

Reply:
Well did you fix chaining in this PR or some other PR?

It doesn't work on the latest version.

Turn on logging for Avaje and I only see it load the top level `load.properties`.

Reply:
it's this pr that'll fix

Reply:
Ok let me pull your branch.

Here is what I think your new chaining is doing:

```properties
# application.properties
load.properties=a.properties last.properties
```

```properties
# a.properties
load.properties=ac.properties
```

```properties
# ac.properties
last=ac
```

```properties
# last.properties
last=last
```


The value of `last` should be `last`. This would only resolve that way if DFS is used.

That is the load order is (by load order I mean the order of when `put` kvs in the mutable config map) should be:

1. application.properties
2. a.properties
3. ac.properties
4. last.properties

But I think what is probably happening and maybe I'm wrong till I pull your code is:

1. application.properties
2. a.properties
3. last.properties
4. ac.properties


I can go into why the above is highly undesirable and I checked our internal config and it does DFS despite me saying the opposite on discord.

```java
// our internal code... notice the insert to the front of the list
// aka stack.
// if it was put on the back it would be a queue and thus BFS.
else if (!subResources.isEmpty()) {
  fs.addAll(0, subResources);
}
```